content/articles/2017-12-14-styler-1-0-0.Rmd

```diff
@@ -0,0 +1,234 @@
+slug: styler-1.0.0
+categories:
+  - package
+photo:
+  url: https://unsplash.com/photos/mpdIPhYqZ4Y
+  author: Heng Films
+description: >
+  styler, a package that can format your code according to the tidyverse style guide, is now on CRAN.
+
+---
+
+```{r setup, include = FALSE}
+knitr::opts_chunk$set(echo = TRUE, comment = "#>")
+knitr::knit_engines$set(list(
+  styler = function(options) {
+    options$comment <- ""
+    knitr::engine_output(
+      options,
+      
+      c("# Before", options$code),
+      c("# After", styler::style_text(options$code))
+    )
+  }
+))
+
+```
+
+
+We're pleased to announce the release of [styler 1.0.0](https://github.com/r-lib/styler/releases/tag/v1.0.0). styler is a source code formatter - a package to format R code according to a style guide. We've implemented the [tidyverse style guide](http://style.tidyverse.org), so you can format your code according to it. A coherent style is important for consistency and legibility. Just as it is important to putspcesbetweenwords. 
+You can install styler from CRAN:
+
+```{r, eval = FALSE}
+install.packages("styler")
+```
+
+styler can style text, single files, packages and entire
+R source trees with the following functions:
+
+* `style_text()` styles a character vector.
+* `style_file()` styles R and Rmd files.
+* `style_dir()` styles all R and/or Rmd files in a directory.
+* `style_pkg()` styles the source files of an R package.
+* An RStudio Addin that styles the active file R or Rmd file, the current 
+  package or the highlighted code.
+
+## styling options
+
+A distinguishing feature of styler is its flexibility. The following options are available:
+
+* `scope`. What is to be styled, i.g. just spacing or also indention, line breaks, tokens?
+* `strict`. How strict should be rules be applied?
+* `math_token_spacing`. Granular control for spacing around math tokens.
+* `reindention`. Allows to set the indention of certain regex patterns manually.
+
+We will briefly describe all of them below.
+
+### scope
+
+`scope` can take one of the following values:
+
+* "none": Performs no transformation at all.
+* "spaces": Manipulates spacing between token on the same line.
+* "indention": In addition to "spaces", this option also manipulates the indention level.
+* "line_breaks": In addition to "indention", this option also manipulates line breaks.
+* "tokens": In addition to "line_breaks", this option also manipulates tokens.
+
+Hence, `scope < "tokens"` will never change code in the sense that the expression remains the same before and after styling, as only spaces and line breaks between the tokens are modified. styler can internally validate the styling, so you can be sure that it has not broken your code. Although extensively tested, we recommend to inspect the styling result with `scope = "tokens"`, since in this case, an automatic validation is not possible.
+
+Let's now see a few examples. We can limit ourselves to styling just spacing information:
```

Comment:
It's conceivable 😮 that chunk options `comment = NA, prompt = TRUE` might make sense here, which I never thought I would say. In this case, it might be easier to appreciate the result if it looked more like code and less like a typical computed result.

Reply:
Good point. We've introduced the styler engine, which works quite nicely here, but for all cases where we also want to show the syntax for the function call (and where the styling options are not `scope = "tokens"` and `strict = TRUE`), we probably can't work around using `style_text()`. But I agree that `comment = NA, prompt = TRUE` would improve legibility.

Reply:
Can we pass options to the styler engine?

Reply:
Probably yes
```r
set_styler_options <- function(options) {
  all_names <- c(
    names(as.list(args(styler::style_text))),
    names(as.list(args(styler::tidyverse_style)))
  )
  options$text <- options$code
  options[intersect(names(options), all_names)]
  
}

knitr::opts_chunk$set(echo = TRUE, comment = NA, promt = TRUE)
knitr::knit_engines$set(list(
  styler = function(options) {
    knitr::engine_output(
      options,
      c("# Before", options$code),
      c("# After", do.call(styler::style_text, set_styler_options(options)))
    )
  }
))
```
So you can pass for example `{styler, scope = "spaces"}` in the header to specify how `# After` looks like.
But then we can't show the syntax anymore, at least not 1:1. Do you know how you could get the options that you specify in the header printed? Even if you can, I think it might be a bit confusing, i.e. 
```
```{styler, scope = "spaces"}
1+11
`` ` 
```

would be the syntax for
```r
style_text("1+11", scope = "spaces")
```

Reply:
What do you think?

Reply:
Looks good!

Reply:
But how can we ensure that the reader does not only see `# Before` and `# After`, but also which options were used and how the syntax looks like? Or do you think it's not necessary to show that? 
In the example above,
```
```{styler, scope = "spaces"}
1+11
`` ` 
```
denotes the code I would insert in the source, but in the rendered html, the reader would only see

```r
# Before
1+11
```
```t
# After
1 + 11
```
which I think does not really convey all the information we want to. So we could either stick with the verbose `style_text()` variant (as it is committed now) or extend the header of the above whenever we don't use the default options, e.g. like this:
```r
# Before
1+11
```
```t
# After style_text(..., scope = "spaces")
1 + 11
```

Reply:
Yeah, I agree it's important to see the syntax in the normal way. My "looks good" was about how the styled code looks. If this was too disruptive, then just go back to the original approach.

Reply:
Ok, I switched to the new syntax with 
```r
# Before
1+11
```


```r
# After style_text(..., [OPTIONS])
1 + 11
```
And added a short explanation of the syntax right after the second code chunk.

Reply:
Ok, I switched to the new syntax with 
```r
# Before
1+11
```


```r
# After style_text(..., [OPTIONS])
1 + 11
```
And added a short explanation of the syntax right after the [second code chunk](https://github.com/tidyverse/tidyverse.org/pull/67/files#diff-5bb48414195a09cf01502ee1fb4c091aR109). See also diff 204d710.

Reply:
Sorry, I missed that. I like the way the `style_text()` call is shown in the comments, would have suggested something similar.